docs/FutureIdeas.md

```diff
@@ -66,3 +66,181 @@ and protocol witness tables, and leave all of this in a context object.
+        }
+        public  MontyWSMBool(): base(MontyWSMBool._XamMontyWSMBoolCtorImpl(), GetSwiftMetatype(), SwiftObjectRegistry.Registry)
+        {
+        }
+        protected  MontyWSMBool(IntPtr handle, SwiftMetatype mt, SwiftObjectRegistry registry): base(handle, mt, registry)
+        {
+        }
+        ~MontyWSMBool()
+        {
+            Dispose(false);
+        }
+        bool BaseVal()
+        {
+            IntPtr thisIntPtr = StructMarshal.RetainSwiftObject(this);
+            bool retval = default(bool);
+            // call into swift to call the base implementation of Val ()
+            retval = NativeMethodsForXam_sub_MontyWSMBool.PImethod_Xam_sub_MontyWSMBoolXamarin_xam_sub_MontyWSMBoolDxam_super_val(thisIntPtr);
+            SwiftCore.Release(thisIntPtr);
+            return retval;
+        }
+        public virtual bool Val()
+        {
+            return BaseVal(); // default implementation, aka, the base.Val
+        }
+#if __IOS__ || __MACOS__ || __TVOS__ || __WATCHOS__
+        [MonoPInvokeCallback(typeof(MontyWSMBool_xam_vtable.Delfunc0))]
+#endif
+		// gets called by swift to implement Val
+        static bool xamVtable_recv_Val(IntPtr self)
+        {
+            return SwiftObjectRegistry.Registry.CSObjectForSwiftObject <MontyWSMBool> (self).Val();
+        }
+        static void XamSetVTable()
+        {
+        }
+        struct MontyWSMBool_xam_vtable
+        {
+            public delegate bool Delfunc0(IntPtr self);
+            [MarshalAs(UnmanagedType.FunctionPtr)]
+            public Delfunc0 func0;
+        }
+    }
+```
+Now you can, in C# override this type and do change the behavior of `Val ()` and if you pass
+an instance of this C# type into Swift, it will pass the handle of the Swift override that
+vectors through your instance.
+
+There is one problem though - when you add new functionality to the the type, Swift will
+know nothing about it. If you have a dozen new C# subclasses of the original Swift class,
+you will not see the new functionality in Swift - they will all look like the singular
+Swift override.
+
+We can fix that - here's one way of doing that.
+Suppose we have an attribute named `[ExposeToSwift]` that can be applied to the type. That way if
+I write this in C#:
+```csharp
+[ExposeToSwift]
+public class Bar : MontyWSMBool {
+}
+```
+Then we make a Roslyn code generator that looks for this and writes the following Swift code:
+```swift
+public class Bar : xam_sub_MontyWSMBool {
+    public override init () { super.init () }
+}
+```
+And call this constructor instead of the `xam_sub_MontyWSMBool` constructor, then Swift will
+now see a class called `Bar` instead of a class called `xam_sub_MontyWSMBool`.
+
+This approach should also work for implementing protocols, including protocols with
+associated types to the degree that the C# type conformance matches any Swift requirements.
+
+In addition, we could allow `[ExposeToSwift]` to be applied to properties and methods as well.
```

Comment:
We do this automatically for Objective-C: if a managed method or property in a subclass overrides an Objective-C method or property, it's exposed automatically.

IMHO this makes sense, because it seems more common to want to expose an overridden Swift method than not.